SSCSM: Allow registering files to be sent, scripts to run from SSM (#17263)
minetest/minetest: doc/sscsm_api.md

Reviewer: Doc needs update.

Reviewer: (Btw, I don't remember if there was any good reason for the `modname:...` path structure. Things might be simpler if we just do `/mods/modname/...`. (Probably not that relevant for this PR though.))

mod with a version check (i.e. at least check if `core.get_version().proto_max`
is (less or) equal to (any of) the tested version(s)).

For security reasons, SSCSM can currently only be used in singleplayer.

> This restriction will be lifted eventually.

Reviewer 2: Under promise over deliver would suggest simply sating that they currently can only be used in singleplayer, but it is true it's on the roadmap.

In SSCSM, the server sends scripts to the client, which it executes
client-side (in a sandbox, see also `sscsm_security.md`).
As modder, you can add these scripts to your server-side mod, and tell the engine
to send them.

Please refer to `lua_api.md` for server-side modding (SSM).
(And refer to `client_lua_api.md` for client-provided client-side modding (CPCSM).)

## Loading mods

### Paths

SSCSM uses a virtual file system (VFS, just a dictionary of virtual paths (strings)
to file contents (strings)).

Each mod's files have paths of the form `modname:foo/bla.lua`.

> Please don't rely on this, use `core.get_modpath()` instead.

Reviewer 2: Do you think perhaps it would be fine to just make this a promise from the API to the developers? This isn't a difficult API constraint to maintain indefinitely, however it does somewhat let mod developers make life harder for themselves if they ever want to rename a mod, if that's the concern it might be helpful to explain that explicitly, "do as I say" is less effective than "this is why you should what I suggest"

> ### Mod sending API

Reviewer: This should probably go into `lua_api.md`. (And here just a reference to it.)

Author: moved it. not sure this is neater but i'm indifferent.

Call `core.register_sscsm(params)` *at load time* from your SSM
`params` is a table with two fields:

> * `init_path`: The virtual relative path of the client mod's init script (equivalent to `init.lua` in SSM)

Reviewer: Should have default, i.e. `client/init.lua`.

Author: I could do that, but `client/init.lua` working requires there being a mapping for it in `paths`. So for just `init_path` to work, I'd need to add a default mapping. But I don't really want to fix/default the mapping too, and if I only fix some paths I would need the ability to disable them.

For now I just want to force mod developers to specify this. If they're lazy, they can just copy from this example.

Reviewer: Fine for now.
I was just worried about mods unnecessarily cooking their own soup, and thereby confusing mod-code readers.

* `paths`: A mapping of virtual relative paths to actual paths in the mod folder

>     * This is a table mapping strings to strings, or `true` if values are to be the same as keys.

Reviewer: `true` should mean that value is relative path, even if key is abs, imo.

Author: hmm? but `true` is the value. the idea is just that i don't want to force modders to spell out `client = "client"` or `["client/init.lua"] = "client/init.lua"` or whatever they choose to map without renaming.

Reviewer: Oops, I think I confused key with value. (And then it would cause issues if you'd take an absolute actual path as virtual path.)

>     * Values can be relative paths of folders, in which case all files will be mapped recursively.

Reviewer: s/of/or/ ?

```suggestion
    * Values can be relative paths.
    * Values can be paths to directories, in which case all files will be mapped recursively.
```

Author: of was intended, as in "path of a folder", though probably should've been "to". will reword anyways.

>     * Values can be relative paths of folders, in which case all files will be mapped recursively.

Reviewer: Should include:
* How are hidden files handled?
* Warn that not only .lua files are transferred. (Privacy concern, and inefficiently transferring crap.)

---

Hm, should there be something to exclude files (i.e. per pattern)?

Author: Hidden files are excluded, will document. Also that file extension doesn't matter. (This I consider a feature, you might want to transfer e.g. `.txt` or `.json` files as part of static resources, if not something fancy like files written in Lua pp langs).

> Hm, should there be something to exclude files (i.e. per pattern)?

maybe in the future, though i'm leaning towards KISS: i would suggest that hidden files / dirs, or separate dirs be used for the purpose of excluding files.

The given files are assumed not to change while the server runs.

Conventionally, client-only scripts should reside in a `client` folder,
including a `client/init.lua` that is run on startup.
Scripts common to client and server (e.g. utilities) are placed in a `common` folder.
Like this:

```lua
core.register_sscsm({
    init_path = "client/init.lua",
    paths = {
       client = true,
       common = true,
    },
})
```

> Init scripts will be run in mod load order.

Reviewer: ```suggestion
Init scripts will be run in SSM mod load order.
```
